base/iterators.jl

```diff
@@ -1095,4 +1094,35 @@ eltype(::Type{Stateful{T, VS}} where VS) where {T} = eltype(T)
 IteratorEltype(::Type{Stateful{T,VS}}) where {T,VS} = IteratorEltype(T)
 length(s::Stateful) = length(s.itr) - s.taken
 
+"""
+    only(x)
+
+Returns the one and only element of collection `x`, and throws an `ArgumentError` if the
+collection has zero or multiple elements.
+"""
+Base.@propagate_inbounds function only(x)
+    i = iterate(x)
+    @boundscheck if i === nothing
+        throw(ArgumentError("Collection is empty, must contain exactly 1 element"))
+    end
+    (ret, state) = i
+    j = iterate(x, state)
+    @boundscheck if j !== nothing
+        throw(ArgumentError("Collection has multiple elements, must contain exactly 1 element"))
+    end
+    return ret
+end
+
+# Collections of known size
+only(x::Tuple{}) = throw(ArgumentError("Tuple is empty, must contain exactly 1 element")
+only(x::Tuple{Any}) = x[1]
+only(x::Tuple) = throw(
+    ArgumentError("Tuple contains $(length(x)) elements, must contain exactly 1 element")
+)
+only(a::AbstractArray{<:Any, 0}) = @inbounds return a[]
```

Comment:
Is it not possible to construct a zero length `AbstractArray{<:Any, 0}` ?

Reply:
How do I test that? i.e. How does one create a zero-dim array of any length?

Reply:
```
julia> collect(1)
0-dimensional Array{Int64,0}:
1
```

Reply:
Idk how to create an empty one

Reply:
I do not know the answer to this (and Andy Ferris is the wise fellow who actually wrote this)